File: ruby/sum-of-multiples/sum_of_multiples.rb
class SumOfMultiples
  def initialize(*multiples)
    @multiples = multiples
  end

  def to(limit)
    (1...limit)
      .select { |number| multiple?(number) }
      .sum
  end

  private

  attr_reader :multiples

  def multiple?(number)
    multiples.any? { |multiple| (number % multiple).zero? }
  end
end

File: typescript/sum-of-multiples/sum-of-multiples.ts
function SumOfMultiples(numbers: number[]) {
  return {
    to(stop: number): number {
      let sum: number = 0

      for (const n of range(1, stop)) {
        if (numbers.some(x => n % x === 0)) {
          sum += n
        }
      }

      return sum
    }
  }
}

function* range(start: number, stop: number) {
  for (let i = start; i < stop; i++) {
    yield i
  }
}

export default SumOfMultiples